src/cargo/sources/git/utils.rs

```diff
@@ -846,32 +883,44 @@ pub fn fetch(
     // The `+` symbol on the refspec means to allow a forced (fast-forward)
     // update which is needed if there is ever a force push that requires a
     // fast-forward.
-    match reference {
-        // For branches and tags we can fetch simply one reference and copy it
-        // locally, no need to fetch other branches/tags.
-        GitReference::Branch(b) => {
-            refspecs.push(format!("+refs/heads/{0}:refs/remotes/origin/{0}", b));
-        }
-        GitReference::Tag(t) => {
-            refspecs.push(format!("+refs/tags/{0}:refs/remotes/origin/tags/{0}", t));
-        }
-
-        GitReference::DefaultBranch => {
-            refspecs.push(String::from("+HEAD:refs/remotes/origin/HEAD"));
-        }
+    if let Some(rev) =
+        locked_rev.filter(|_| !matches!(shallow, gix::remote::fetch::Shallow::NoChange))
+    {
+        // If we want a specific revision and know about, obtain that specifically.
+        refspecs.push(format!("+{0}:refs/remotes/origin/HEAD", rev));
+    } else {
+        match reference {
+            // For branches and tags we can fetch simply one reference and copy it
+            // locally, no need to fetch other branches/tags.
+            GitReference::Branch(b) => {
+                refspecs.push(format!("+refs/heads/{0}:refs/remotes/origin/{0}", b));
+            }
+            GitReference::Tag(t) => {
+                refspecs.push(format!("+refs/tags/{0}:refs/remotes/origin/tags/{0}", t));
+            }
 
-        GitReference::Rev(rev) => {
-            if rev.starts_with("refs/") {
-                refspecs.push(format!("+{0}:{0}", rev));
-            } else if let Some(oid_to_fetch) = oid_to_fetch {
-                refspecs.push(format!("+{0}:refs/commit/{0}", oid_to_fetch));
-            } else {
-                // We don't know what the rev will point to. To handle this
-                // situation we fetch all branches and tags, and then we pray
-                // it's somewhere in there.
-                refspecs.push(String::from("+refs/heads/*:refs/remotes/origin/*"));
+            GitReference::DefaultBranch => {
                 refspecs.push(String::from("+HEAD:refs/remotes/origin/HEAD"));
-                tags = true;
+            }
+
+            GitReference::Rev(rev) => {
+                if rev.starts_with("refs/") {
+                    refspecs.push(format!("+{0}:{0}", rev));
+                } else if let Some(oid_to_fetch) = oid_to_fetch {
+                    refspecs.push(format!("+{0}:refs/commit/{0}", oid_to_fetch));
+                } else if is_shallow && git2::Oid::from_str(&rev).is_ok() {
+                    // There is a specific commit to fetch and we will just do so in shallow-mode only
+                    // to not disturb the previous logic. Note that with typical settings for shallowing,
+                    // we will just fetch a specific slice of the history.
+                    refspecs.push(format!("+{0}:refs/remotes/origin/HEAD", rev));
```

Comment:
Maybe we could check whether it is a full hash to determine what to do. However, 40 characters hex strings might not be a full hash on if git server use SHA256. See https://github.com/rust-lang/cargo/pull/10807

Reply:
> It works, but I don't really understand. Could you explain why fetching `refs/remotes/origin/HEAD` to `rev` works for fetching “a slice of history”, and why we need to treat shallow differently?


We fetch `rev` which is a full object hash (this is important, as only full object hashes are allowed in refspecs) and set `refs/remotes/origin/HEAD` to point to it. This ref name [is used specifically](https://github.com/rust-lang/cargo/blob/393ce5b41632173199d3b996de9c8a89f6290435/src/cargo/sources/git/utils.rs#L217) by `cargo` to resolve `GitReferences`, and at least in our test-suite that's happening often enough.

Furthermore, I have re-enabled the fast-path in 493574eb2ed6c3115f53448bc00a3bf6eea2cbc9 so in the hopefully rare case of somebody using a partial object name it will be resolved if GitHub is used.

Besides that, object prefixes/short hashes are invalid in that position and would ideally be rejected by `cargo` earlier. Unfortunately, I don't think this can be done as it also prevents people from referring to `deadbeef` objects.

Reply:
Really appreciate the explanation!

> object prefixes/short hashes are invalid in that position and would ideally be rejected by cargo earlier. 

Hmm… this is not exactly true I feel like. The fetch refspecs previously fell back to `+refs/heads/*:refs/remotes/origin/*` and `+HEAD:refs/remotes/origin/HEAD` if a dep is specified with a `rev` and Cargo cannot resolve it. Now the fallback is gone. The path is pretty rare as Cargo can almost always get a full object hash through GitHub fast path, though it's still a regression for repos not on GitHub, preventing some projects from building.

Reply:
I wonder if we want to check if it is a full object hash for this branch, or have a more robust retry mechanism, or make this branch shallow-only temporarily and deal with it later on.

Reply:
I might have overshot by removing the `shallow-only` condition, and also feel more comfortable only use this branch for where it is actually needed. The blanket 'fetch-all' handling seems fine and changes this even if only a single-commit is needed might be a breaking change. 

However, I don't know how such breakage would work - the `dep_with_bad_submodule` test failed because the submodule wasn't actually broken anymore, for example, which seems that avoiding blanket fetches makes it more robust. 

Since my 'try harder to break the submodule' fix also [broke windows](https://github.com/rust-lang/cargo/actions/runs/4838271418/jobs/8622632953?pr=11840#step:11:3423), I didn't think long to bring back the `is_shallow` condition though, here in 2aaebcbcaa66b799a807ba67d1ee7b6b44f7bdc7 .

What do you think about this?

Reply:
Looks good! We still have room to optimize with fallbacks and retries. I am not too concerned with that at this moment, but perhaps we mark it as unresolved in the tracking issue.